bin/dump_data_to_file.py

```diff
@@ -0,0 +1,206 @@
+import json
+import os
+import math
+import requests
+import time
+import arrow
+import argparse
+import sys; sys.path.append("..")
+from emeval.input.spec_details import ServerSpecDetails
+from emeval.input.phone_view import PhoneView
+
+
+def dump_data_to_file(data, spec_id, user, key, start_ts, end_ts, out_dir):
+    """
+    Accepts serializable data (e.g. dict, array of dicts) and dumps it into an output file.
+    Dumped file are created recursively in the folder name specified by `out_dir` as such:
+
+    out_dir
+    └── user
+        └── spec_id
+            └── key
+                └── {start_ts}_{end_ts}.json
+    """
+    # key could have a slash if it corresponds to a key_list argument in SpecDetails::retrieve_data
+    # slashes are invalid in directory names, so replace with tilde
+    out_path = os.path.join(out_dir, user, spec_id, key.replace("/", "~"))
+    os.makedirs(out_path, exist_ok=True)
+
+    out_file = os.path.join(out_path, f"{math.floor(start_ts)}_{math.ceil(end_ts)}.json")
+    print(f"Creating {out_file=}...")
+    with open(out_file, "w") as f:
+        json.dump(data, f, indent=4)
+
+
+def make_call_to_server(datastore_url, author_email, user, key, start_ts, end_ts):
+    """
+    Makes a direct call to the E-Mission Server instance based on the specified user/key/start_ts/end_ts.
+    """
+    return ServerSpecDetails(datastore_url, author_email).retrieve_one_batch(user, [key], start_ts, end_ts)
+
+
+def get_all_spec_ids(datastore_url, author_email):
+    """
+    Retrieves list of all spec_id's on E-Mission Server instance being used by script.
+    """
+    spec_data = make_call_to_server(
+        datastore_url,
+        author_email,
+        author_email,
+        "config/evaluation_spec",
+        0,
+        sys.maxsize)
+
+    spec_ids = [s["data"]["label"]["id"] for s in spec_data]
+
+    return set(spec_ids)
+
+
+def run_full_pipeline(datastore_url, author_email, spec_ids, out_dir):
+    """
+    Runs the full data retrieval pipeline in the event that a user/key/start_ts/end_ts combination isn't provided.
+    """
+    print(f"Running full pipeline for {spec_ids[0] if len(spec_ids) == 1 else 'all specs in datastore'}...")
+
+    # collect ServerSpecDetails objects, dump specs
+    sds = []
+    for s_id in spec_ids:
+        sd = ServerSpecDetails(datastore_url, author_email, s_id)
+        sds.append(sd)
+        dump_data_to_file(
+            sd.curr_spec_entry,
+            sd.CURR_SPEC_ID,
+            author_email,
+            "config/evaluation_spec",
+            0,
+            sys.maxsize,
```

Comment:
Why is this `sys.maxsize` again?

Reply:
The issue I was running into here was that the `end_ts` of the spec was equal to `arrow.get().timestamp`, which obviously changes each time you try to retrieve a spec. `sys.maxsize` was the best convention I could think of for representing an all-inclusive time range, as due to it's constant nature it ensures valid lookups for spec details.